### ISSUE TITLE
pos[1] json: expect char '"' but got char '{'

### ISSUE THREAD

**Issue by caesarxuchao** [ugorji/go#178]
I'm trying to let the ugorji codec to call the UnmarshalText() method I defined for my type, but the ugorji codec panic before it calls `UnmarshalText()`.

The call to ugorji ([link](https://github.com/kubernetes/kubernetes/blob/v1.4.5/pkg/runtime/serializer/json/json.go#L124)):
```go
codec.NewDecoderBytes(data, new(codec.JsonHandle)).Decode(into); err != nil
```

where data is:
```
{"kind":"ExampleList","items":[{"apiVersion":"wfarr.systems/v1","kind":"Example","metadata":{"name":"example1","namespace":"default","selfLink":"/apis/wfarr.systems/v1/namespaces/default/examples/example1","uid":"9fa56da5-9ef1-11e6-b22c-64510650ab1c","resourceVersion":"89932","creationTimestamp":"2016-10-30T22:38:48Z"},"spec":{"bar":true,"foo":"Hello"}}],"metadata":{"selfLink":"/apis/wfarr.systems/v1/examples","resourceVersion":"102516"},"apiVersion":"wfarr.systems/v1"}
```

`ExampleList` is defined [here](https://github.com/caesarxuchao/k8s-tpr-playground/blob/master/main.go#L26-L38). Its `UnmashalText()` is defined this way:
```go
func (el *ExampleList) UnmarshalText(data []byte) error {
	return json.Unmarshal(data, el)
}
```

And ugorji codec failed at this [line](https://github.com/ugorji/go/blob/master/codec/decode.go#L486):
```go
tm.UnmarshalText(f.d.d.DecodeBytes(f.d.b[:], true, true))
```

Because it panic at this [line](https://github.com/ugorji/go/blob/master/codec/json.go#L972):
```go
d.d.errorf("json: expect char '%c' but got char '%c'", '"', d.tok)
```

@ugorji could take a look? It seems like a bug, but it could be my understanding of the codec is wrong.

**Comment by ugorji** [comment]
Hi,

It's hard to understand and resolve an issue without a straightforward reproducer.

Can you please provide
- sample file.go or test_file.go that I can run with either go run or go test
- clarify what you expect to see on execution, and what you see instead (and why that is wrong)

With those, I can explain the behavior, or reproduce and fix.

**Comment by ugorji** [comment]
@caesarxuchao ^^

**Comment by caesarxuchao** [comment]
Sure. I'll do it later today. Thank you.

**Comment by caesarxuchao** [comment]
Hi @ugorji, sorry I didn't have time update on this issue during the week. I created an example here: https://github.com/caesarxuchao/ugorji-reproduce-error. Thanks for the help!

**Comment by ugorji** [comment]
Aha - this is a User Error.

TextMarshaler and JSONMarshaler (and the unmarshal variants) are subtly different.

JSON will wrap up the value in *""*, and will expect the string wrapped in *""* when decoding. However, TextMarshaler is different.

As an example, look at time.go
- JSONMarshal will produce *"2006-01-02T15:04:05Z07:00"*
- TextMarshal will produce *2006-01-02T15:04:05Z07:00* (notice the missing quotes)

Consequently, JSONUnmarshal will expect *"2006-01-02T15:04:05Z07:00"* while UnmarshalText will expect *2006-01-02T15:04:05Z07:00*

In summary, it is an error to use json package to implement Marshal or Unmarshal when the underlying type is a string, except you deliberately handle that.

See https://golang.org/src/time/time.go?s=27433:27480#L934 for how time.Time handles marshal/unmarshal differently for TextMarshaler vs JSONMarshaler. In general, JSONMarshaler should be json.Marshal(string(x.MarshalText())) or something like that.

**Comment by caesarxuchao** [comment]
Thanks for the explanation. But why the panic message says `expect char '"' but got char '{'`? The input of TextUnmarshal should not start with `"`.

**Comment by caesarxuchao** [comment]
Also, given that I cannot define TextUnmarshal(), is there a clean way to use instruct ugorji to use the golang json unmarshaller for a type? Thanks.
